apache/superset
# chore: Convert QueryAutoRefresh to TypeScript functional React component [sc-48362]

Convert QueryAutoRefresh to TypeScript functional React component

### SUMMARY

This PR converts QueryAutoRefresh from es6 class component to TypeScript functional component as part of bigger effort to migrate frontend code to TypeScript and functional React components defined in: https://github.com/apache/superset/discussions/18388

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

There should be no visual change to the application

### TESTING INSTRUCTIONS

1. Open SQL Lab / SQL Editor
2. Click the "RUN" button
3. Ensure Query runs and updates results with same behavior that existed prior to this PR
4. Test with both default behavior and asynchronous query execution on for a database

**Changes to http requests for getting Query statuses:**
If you watch the dev tools network tab you should see less network traffic as the execution now ensures only one outstanding http call to check query status is made at a time to avoid http response race conditions that can lead to incorrect application state and to reduce tandem network calls on slower connections.  If the call fails, the offline state change will be triggered but the interval timer will continue attempting to reconnect and set user online and show results once the connection is restored.  This can be tested by:
1. Opening the dev tools / Network / set the connection from "No Throttling" to "Offline", 2. Click the "Run" button as described in the initial test steps above
3. See the network failure and the UI will display an "Offline" pill (this is not new UI behavior)
4. Set the network back to "No Throttling" and wait a few seconds and the next interval will make a successful query status API call and you will see data results again.

QueryAutoRefresh was separated into two files to separate Redux knowledge from the simple functional component:
1. QueryAutoRefresh/index.ts - Simple functional component with no knowledge of Redux
3. QueryAutoRefresh/QueryAutoRefreshContainer.ts - Functional component container with connection to redux state that imports and contains QueryAutoRefresh/index.ts component

**IMPORTANT:** 
When changing QueryAutoRefresh to a functional component the behavior of setInterval no longer worked correctly.  This is a known issue with functional React components using setInterval can provide stale state / props in the interval callback creating undesired, hard to debug / fix behavior.  A new utility hook (SqlLab/utils/useInterval.ts) was introduced in this PR to handle this based on this very helpful article by Dan Abramov (thank you!): https://overreacted.io/making-setinterval-declarative-with-react-hooks/

### ADDITIONAL INFORMATION


- [X] Has associated issue: https://github.com/apache/superset/discussions/18388
